package/libpiksi/libpiksi/test/test_version.cc

```diff
@@ -93,10 +93,16 @@ TEST_F(LibpiksiTests, versionTests)
     EXPECT_EQ(1, ver.major);
     EXPECT_EQ(4, ver.patch);
 
+    EXPECT_EQ(0, version_parse_str("v2.0.0-develop-2018101616-11-g8a", &ver));
+    EXPECT_EQ(2, ver.marketing);
+    EXPECT_EQ(0, ver.major);
+    EXPECT_EQ(0, ver.patch);

```

Comment:
Won't this prevent users from upgrading between different dev versions with the same base versions?  I'm wondering if version comparison can simply be a string comparison?

Reply:
I think the reasoning why I initially drifted away from a basic string comparison was that I didn't see any official format nomenclature for the version string. I have seen version with/without leading `v` for example.

And how about a case `2.0.2` vs `2.0.10`?

Reply:
But yes your comment about dev versions with same base is correct.

Reply:
I'll think about this a bit if anything sensible can be done.

Reply:
ok, how about this, if `-<devstring>` is detected in the offered version, upgrade is always performed? See latest implementation.

Reply:
I don't think continually upgrading in the presence of a dev image is a good idea unless we are saying "only upgrade if dev AND it's a different dev image" -- otherwise this puts onus on the server to stop sending the new update version (minimizing logic on the server side, and client side is a good thing for OTA).

Reply:
That makes sense.

Reply:
Sorry, one final comment about version comparison: the OTA daemon needs to allow version downgrades, if for some reason we need to roll back a bunch of devices on field (for using Duro as a base station) this will be a critical capability.  I think it's good to have the parsing and logging on the "directionality" of the upgrade, but the logic at this point is really just "if the offered version is different than the current version, upgrade".

Reply:
No problem, it's better to do this now than after we already shipped.

I reworked the `ota_version_check` logic.
Upgrade/rollback happens if:
* Development build replaces production build (base version ignored)
* Production build replaces development build (base version ignored)
* Development build with different devstring replaces development build (base version ignored)
* Newer production build replaces older production build
* Older production build replaces newer production build

Reply:
One thing I am still worried. Currently the leading chars before first digit are ignored.. Need to think about that for a moment.

Reply:
OK, I took the possible leading chars as part of devstring. Logic is getting complicated trying to cover all the cases though.. I think it's borderline acceptable still, but anything more and I think reverting back to just `strcmp` is all the time more tempting option.